### PR TITLE
Insert content next to widgets

### LINKED ISSUE #4467: Inserting content next to block widgets using keyboard navigation
## Type of report

Feature request

## Provide description of the new feature

Currently, if two block widgets are inserted right after each other, there is no possibility to add content between them without either some copy&paste workarounds or [Magic Line plugin](https://ckeditor.com/cke4/addon/magicline). Could be useful if it was possible in some friendly way.

Since the issue only concerns block widgets, expected result would probably be inserting the paragraph after the widget (like with Magic Line, but without mouse interaction). The `enter` keystroke is already reserved for widget edition dialog, so we need another one. A possibility is `shift+enter`, which is currently doing nothing in such scenario (it should be investigated if `enterkey` plugin doesn't alter this behaviour though).

That part would solve the problem with adding content **between** two widgets or **after** the widget if it is placed at the end of the document. There is one more situation to consider which is a widget at the beginning of the document - we would need another keystroke for inserting the paragraph **before** the widget, but that may be an edge case worth further extracting.

For more context, there is https://github.com/ckeditor/ckeditor4/pull/4433 which probably enables removing the existing paragraph from between two widgets, which may cause the situation with two "glued" widgets more often, so it would be good to have it introduced in the same release as this feature.

### PR DESCRIPTION
**Pull request by bunglegrind**
## What is the purpose of this pull request?

New feature. See #4467 

## Does your PR contain necessary tests?

Yes.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [X] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4467](https://github.com/ckeditor/ckeditor4/issues/4467): Insert a new paragraph next to a widget.
```

## What changes did you make?

See https://github.com/ckeditor/ckeditor4/issues/4467#issuecomment-765311509 

## Which issues does your PR resolve?

Closes #4467.

### CODE REVIEW

File: tests/plugins/widget/manual/insertparagraph.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+@bender-tags: widget, faeture, 4.16.0, 4467
```
Comment (Dumluregn) [inline tests/plugins/widget/manual/insertparagraph.md:1]: Please bump version tag.

**Review by Dumluregn — CHANGES_REQUESTED** [review]
PR is already looking very good 👍🏻  Fix works on all browsers and is compatible with changes introduced in #4433. Big 👍🏻  for keeping the code style very coherent with CKE4 standards.

I've pointed out only a few small things.

File: plugins/widget/plugin.js [hunk 2]
```diff
@@ -3461,6 +3461,19 @@
 		}
 	} );
 
+	function insertParagraphAndSaveSnapshot( widget, where ) {
+		var p = new CKEDITOR.dom.element( 'p' );
+
+		p.appendBogus();
+
+		if ( where === 'below' ) {
+			widget.wrapper.$.parentNode.insertBefore( p.$, widget.wrapper.$.nextSibling );
```
[the remark this answers is not in the document]
Comment (bunglegrind) [reply plugins/widget/plugin.js:3470]: fixed

File: plugins/widget/plugin.js [hunk 3]
```diff
@@ -3461,6 +3461,19 @@
...
+		} else {
+			p.insertBefore( widget.wrapper );
+		}
```
[the remark this answers is not in the document]
Comment (bunglegrind) [reply plugins/widget/plugin.js:3473]: yes! Thanks for the suggestion!

File: tests/plugins/widget/manual/insertparagraph.md (hunk 1 above)
Comment (bunglegrind) [reply tests/plugins/widget/manual/insertparagraph.md:1]: Increased to 4.16.1. I don't know if you're using semantic versioning...

File: plugins/widget/plugin.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (bunglegrind) [reply plugins/widget/plugin.js:3470]: Ops.

**Comment by bunglegrind** [comment]
Failing...because: https://github.com/ckeditor/ckeditor4/issues/3868

**Comment by Dumluregn** [comment]
> Failing...because: #3868

Yes, all branches started to fail because of this test, so we can ignore it in this PR. Otherwise ready for review I suppose?

**Comment by bunglegrind** [comment]
yes, ready!

**Review by Dumluregn — APPROVED** [review]
I've added handling for different [`config.enterMode`](https://ckeditor.com/docs/ckeditor4/latest/examples/enterkey.html) settings, API docs for the new config options and a bunch of tests.

@ckeditor/qa-team, please take a look if we didn't break anything in the process.

**Comment by Dumluregn** [comment]
I've also created an issue for keyboard shortcuts docs update - https://github.com/ckeditor/ckeditor4/issues/4559.

**Comment by LukaszGudel** [comment]
Hi, I've looked into this feature and everything seems to work correctly :+1:

**Comment by Dumluregn** [comment]
Ok, so my last doubt is if we should merge it to `master` or `major` branch 🤔  Since it introduces a feature, it should probably be targeted at `major`, but it is tightly connected with #1572, which was merged into `master`. So @f1ames please say WDYT about it.

**Comment by f1ames** [comment]
> Ok, so my last doubt is if we should merge it to `master` or `major` branch Since it introduces a feature, it should probably be targeted at `major`, but it is tightly connected with #1572, which was merged into `master`. So @f1ames please say WDYT about it.

Well, it is connected with #1572 indeed but OTOH it can work separately and does not depend on fixes provided in #1572. AFAIU it makes sense even without #1572 (correct me if I'm wrong @Dumluregn) so I would be for merging it according to our standard set of rules (to `master`).

And then when scheduling next release we can think if it makes sense to go with `minor` or we should go with `major` which includes both changes.

**Comment by Dumluregn** [comment]
Rebased onto `major` branch, waiting for CI to merge it.

**Comment by Dumluregn** [comment]
We only have a failing test in https://github.com/ckeditor/ckeditor4/issues/3868, so I'm merging this PR - thanks again for the contribution @bunglegrind !

**Pull request merged**